content/articles/react-query-data-fetching-and-server-state-management/index.md

```diff
@@ -0,0 +1,255 @@
+- Inside the JSX part, we return the posts by mapping through the posts array.
+
+> For an awesome developer experience, `react-query` includes the in-built `react-query-devtools`, so that we will be able to view the state and the cache.
+
+To enable this in your application `import { ReactQueryDevtools } from 'react-query/devtools'`.
+
+You can read more about this [here](https://react-query.tanstack.com/devtools).
+
+### Making a POST request using `useMutation` hook
+Unlike the `useQuery` hook, mutations are used to perform `CREATE`, `UPDATE`, or `DELETE` operations as the server side operations inside our components.
+
+In our case, we will perform a `POST` request using `useMutation`.
+
+Inside the `components` folder, create a `Post.js` file.
+
+In the `Post.js` component file, import `axios`, `React` and `useMutation` hook as shown:
+
+```js
+import React, {Fragment, useState} from 'react'
+import { useMutation } from 'react-query'
+
+import axios from 'axios'
+```
+
+Next, create a `Post` functional component:
+
+```js
+export default function Post(){
+    const [title, setTitle] = useState('')
+    const [description, setDescription] = useState('')
+    const [message, setMessage] = useState('')
+
+    const {isLoading, isError, error, mutate} = useMutation(createPost, {retry: 3})
+
+
+    async function createPost() {
+        const response = await axios.post('https://jsonplaceholder.typicode.com/posts')
+        setMessage(response.data)
+    }
+
+    return(
+        <Fragment>
+            <div className="post">
+            <h1>Create a Post</h1>
+
+                <label>Title:</label>
+                <input type="text" value={title} onChange={e=>setTitle(e.target.value)}/>
+
+                <label>Description:</label>
+                <input type="text" value={description} onChange={e=>setDescription(e.target.value)}/>
+                <button  onClick={() => {mutate({ id: Date.now(), title, description })}} >Create</button>
+                <p> Created a new Post ID: {message && message.id}</p>
+                <div style={{color: 'gray', background: '#234'}}>
+        {isLoading
+          ? "Saving...": ""
+        }
+        {
+            isError
+            ? error.message : ""
+        }
+      </div>
+      </div>
+
+        </Fragment>
+    )
+}
+```
+
+To briefly explain the above code:
+
+- `createPost` is a function that makes HTTP POST request using `axios` library.
+- The `useMutation()` hook returns `isLoading`, `isError`, `error`, and `mutate` function that will be used to wrap the values when making requests. It takes the `createPost` as an argument along with an option `{retry:3}` object. This comes in handy when prefetching queries after mutations and ensuring optimistic updates.
+- `useState()` hook is used to create and update the `title`, and description state in our input elements.
+- Below the create post functionality, we use the `isLoading` and `isError` to handle the mutation state accordingly.
+
```

Comment:
This has an abrupt ending. The reader will not be able to conclude anything from here.